src/MessageWriter.ts

```diff
@@ -373,112 +407,162 @@ function uint8Array(value: unknown, defaultValue: DefaultValue, writer: CdrWrite
+  defaultValue: DefaultValue,
+  writer: CdrWriter,
+  arrayLength?: number,
+): void {
   if (value instanceof BigInt64Array) {
     writer.int64Array(value);
   } else if (Array.isArray(value)) {
     const array = new BigInt64Array(value);
     writer.int64Array(array);
   } else {
-    writer.int64Array((defaultValue ?? []) as bigint[]);
+    writer.int64Array((defaultValue ?? new Array(arrayLength).fill(0n)) as bigint[]);
   }
 }
 
-function uint64Array(value: unknown, defaultValue: DefaultValue, writer: CdrWriter): void {
+function uint64Array(
+  value: unknown,
+  defaultValue: DefaultValue,
+  writer: CdrWriter,
+  arrayLength?: number,
+): void {
   if (value instanceof BigUint64Array) {
     writer.uint64Array(value);
   } else if (Array.isArray(value)) {
     const array = new BigUint64Array(value);
     writer.uint64Array(array);
   } else {
-    writer.uint64Array((defaultValue ?? []) as bigint[]);
+    writer.uint64Array((defaultValue ?? new Array(arrayLength).fill(0n)) as bigint[]);
   }
 }
 
-function float32Array(value: unknown, defaultValue: DefaultValue, writer: CdrWriter): void {
+function float32Array(
+  value: unknown,
+  defaultValue: DefaultValue,
+  writer: CdrWriter,
+  arrayLength?: number,
+): void {
   if (value instanceof Float32Array) {
     writer.float32Array(value);
   } else if (Array.isArray(value)) {
     const array = new Float32Array(value);
     writer.float32Array(array);
   } else {
-    writer.float32Array((defaultValue ?? []) as number[]);
+    writer.float32Array((defaultValue ?? new Array(arrayLength).fill(0)) as number[]);
   }
 }
 
-function float64Array(value: unknown, defaultValue: DefaultValue, writer: CdrWriter): void {
+function float64Array(
+  value: unknown,
+  defaultValue: DefaultValue,
+  writer: CdrWriter,
+  arrayLength?: number,
+): void {
   if (value instanceof Float64Array) {
     writer.float64Array(value);
   } else if (Array.isArray(value)) {
     const array = new Float64Array(value);
     writer.float64Array(array);
   } else {
-    writer.float64Array((defaultValue ?? []) as number[]);
+    writer.float64Array((defaultValue ?? new Array(arrayLength).fill(0)) as number[]);
   }
 }
 
-function stringArray(value: unknown, defaultValue: DefaultValue, writer: CdrWriter): void {
+function stringArray(
+  value: unknown,
+  defaultValue: DefaultValue,
+  writer: CdrWriter,
+  arrayLength?: number,
+): void {
   if (Array.isArray(value)) {
     for (const item of value) {
       writer.string(typeof item === "string" ? item : "");
     }
   } else {
-    const array = (defaultValue ?? []) as string[];
+    const array = (defaultValue ?? new Array(arrayLength).fill("")) as string[];
     for (const item of array) {
       writer.string(item);
     }
   }
 }
 
-function timeArray(value: unknown, _defaultValue: DefaultValue, writer: CdrWriter): void {
+function timeArray(
+  value: unknown,
+  _defaultValue: DefaultValue,
+  writer: CdrWriter,
+  _arrayLength?: number,
+): void {
```

Comment:
why no defaultValue fallback here? Seems like we would need it for time arrays as well... or is there no such thing as a fixed-length time array?